### ISSUE TITLE
Remove the usage of `data` blocks for resources created in the budget module

### ISSUE THREAD

**Issue by uplight-james** [terraform-google-modules/terraform-google-project-factory#802]
### TL;DR

The [data block here](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/modules/budget/main.tf#L35) will always [refer to the project created by the same module call](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/v14.2.0/main.tf#L92) unless you call budget as a submodule.

The usage of `depends_on` in that datablock will typically [defer reading of that data block until apply time](https://developer.hashicorp.com/terraform/language/data-sources#data-resource-dependencies), which makes plans less accurate than they could be for seemingly little gain here.

### Expected behavior

The first time you create a budget, you will see computed/`known-after-apply` values, subsequent applications should not display messages about deferring reads until apply and should instead just show an actual plan (which itself may or may not show computed values based on what changed). 

Changes to the project created in the same module should also not cause deferred reading of data blocks.

### Observed behavior

The current behavior means that any changes to how we call the project factory that change the `google_project` created in the core_project_factory module result in this output occurring:

```
  # module.<signature> module.budget.data.google_project.project[0] will be read during apply
  # (config refers to values not yet known)
 <= data "google_project" "project"  {
      + auto_create_network = (known after apply)
      ~ billing_account     = "012345-123456-123456" -> (known after apply)
      ~ folder_id           = "1234567890123" -> (known after apply)
      ~ id                  = "projects/my-project-id-123c" -> (known after apply)
      ~ labels              = {
          - "label1"                 = "value1"
          - "label2"               = "value2"
          - "label3"                   = "value3"
        } -> (known after apply)
      ~ name                = "my-project-9d" -> (known after apply)
      ~ number              = "1234567890123" -> (known after apply)
      + org_id              = (known after apply)
      + skip_delete         = (known after apply)
        # (1 unchanged attribute hidden)
    }
```

If you miss the `will be read` part, this can be kind of concerning as you see a lot of `-` and `~` signs for things you might not have changed. Obviously in this case its terraform just recognizing that it can't know what those values will ultimately be but it's still a bit concerning right away, particularly if you miss the `will be read` part.

Additionally, hashicorp itself doesn't recommend this unless you really have to (see my link to `depends_on` above) as if you make determinations on things to change based on `data` changes you might obviously have changes that you didn't realize wouldn't occur. 

In this case, thats not super likely to happen but it is a bit of a code smell. And as its concerning, it requires users to dig down into the code to try to forecast what about this module ACTUALLY changes.

Ultimately this just gives a slightly worse signal-to-noise ratio.

### Terraform Configuration

```hcl
module "my_project" {
  source                  = "terraform-google-modules/project-factory/google"
  version                 = "~> 13.0"
  random_project_id       = false
  default_service_account = "deprivilege"
  activate_apis           = ["logging.googleapis.com",
  name                    = "my-project-name"
  org_id                  = local.org_id
  billing_account         = local.billing_account
  folder_id               = local.folder_id
  lien                    = local.create_lien
  svpc_host_project_id    = "<Our svpc project id>"
  grant_network_role      = var.grant_network_role

  labels = local.new_labels
}
```

You should be able to observe this behavior by:

1. Plan the project - you will see the `will be read` message
2. Apply the project
3. Change a variable that would cause a change to the project in the [core-project-factory](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/modules/core_project_factory/variables.tf) such as adding/changing a label, execute `terraform plan`

Because references to that resource are fed into budget and because you use `depends_on`, terraform defers reading and gives you the message above. Note that in this case, you didn't even change anything that modified your budget.
```


### Terraform Version

```sh
Terraform v1.1.9
on darwin_amd64
+ provider registry.terraform.io/datadog/datadog v3.22.0
+ provider registry.terraform.io/hashicorp/external v2.3.1
+ provider registry.terraform.io/hashicorp/google v4.57.0
+ provider registry.terraform.io/hashicorp/google-beta v4.57.0
+ provider registry.terraform.io/hashicorp/kubernetes v1.13.4
+ provider registry.terraform.io/hashicorp/local v2.4.0
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/random v3.4.3
+ provider registry.terraform.io/hashicorp/time v0.9.1
```

### Additional information

It looks like all you are using this block for is to pull up the [`project number`](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/v10.3.3/modules/budget/main.tf#L24). Admittedly its a tradeoff but you could change the `project` variables to accept a `list(object)`:

```
variable "project" {
   type = list(object({
      project_id           = string
      project_number = number
   }))
}
```

Thats admittedly a bit clunky but it does remove the need to defer lookups. 

Also I am currently sourcing V13.x of this module but I believe it affects newer versions as well.

**Comment by uplight-james** [comment]
I think this is still an issue.
